### PR TITLE
[mlir][Vector] Fix bug in vector xfer op flattening transformation

### PR DESCRIPTION
**Pull request by dcaballe**
It looks like the affine map generated to compute the indices of the collapsed dimensions used the wrong dim size. For indices `[idx0][idx1]` we computed the collapsed index as `idx0*size0 + idx1` instead of `idx0*size1 + idx1`. This led to correctness issues in convolution tests when enabling this transformation internally.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp [hunk 1]
```diff
@@ -615,10 +615,14 @@ class FlattenContiguousRowMajorTransferReadPattern
...
+      auto srcType = dyn_cast<ShapedType>(source.getType());
       for (int64_t i = firstDimToCollapse; i < outputRank; ++i) {
-        int64_t dim = dyn_cast<ShapedType>(source.getType()).getDimSize(i);
+        // Multiply each index by the size of the next dimension. The last
+        // dimension (contiguous) is multiplied by one.
+        int64_t nextDimSize =
+            (i == outputRank - 1) ? 1 : srcType.getDimSize(i + 1);
         offset = affine::makeComposedFoldedAffineApply(
-            rewriter, loc, offsetExpr + dim * idxExpr,
+            rewriter, loc, offsetExpr + nextDimSize * idxExpr,
```
Comment (MacDue) [inline mlir/lib/Dialect/Vector/Transforms/VectorTransferOpTransforms.cpp:625]: Just double checking is this correct for > 2D? 

`[idx0][idx1][idx2]` would be:

 `idx0 * size1*size2 + idx1 * size2 + idx2`

File: mlir/test/Dialect/Vector/vector-transfer-flatten.mlir [hunk 2]
```diff
@@ -66,14 +66,14 @@ func.func @transfer_read_dims_mismatch_non_zero_indices(
...
-  %2 = vector.transfer_read %m_in[%c0, %idx_1, %idx_2, %c0], %c0_i32 {in_bounds = [true, true, true]} : 
+  %2 = vector.transfer_read %m_in[%c0, %idx_1, %idx_2, %c0], %c0_i32 {in_bounds = [true, true, true]} :
     memref<1x43x4x6xi32>, vector<1x2x6xi32>
   vector.transfer_write %2, %m_out[%c0, %c0, %c0] {in_bounds = [true, true, true]} :
     vector<1x2x6xi32>, memref<1x2x6xi32>
   return
 }
 
-// CHECK: #[[$ATTR_0:.+]] = affine_map<()[s0, s1] -> (s0 * 4 + s1 * 43)>
+// CHECK: #[[$ATTR_0:.+]] = affine_map<()[s0, s1] -> (s0 * 6 + s1 * 4)>
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:76]: Shouldn't this be `s0 * 6 + s1 * (4 * 6)`? As in, this affine map looks incorrect to me.

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]
We have IndexingUtils to avoid reinventing such code every time..

Can we please use that and improve / evolve as needed?

**Comment by dcaballe** [comment]
Thanks! Good catch! Using index utils now. It should ready

**Review by hanhanW — APPROVED** [review]
LGTM, thanks!

**Comment by dcaballe** [comment]
Thanks! Will land this tomorrow if no more comments.

**Pull request merged**